### PR TITLE
Introduce MultiSourceSymbolMapping

### PR DESCRIPTION
**Pull request by kokosing**
Introduce MultiSourceSymbolMapping

MultiSourceSymbolMapping contains list of source nodes, output symbols
and mapping of input symbols (the ones which come from source nodes) to
output symbols.

Motivation behind this is to use well defined entities instead of
complex collections.

The same thing exists in ExchangeNode, I will try to use
MultiSourceSymbolMapping there as a follow up of this.

### CODE REVIEW

**Comment by kokosing** [comment]
@sopel39 Can you please take a look. I am curious of your opinion about this before moving any further. The next step is to use it in `ExchangeNode` where there is the same concept but it is handled totally different. The goal is to unify them.

File: presto-main/src/test/java/com/facebook/presto/execution/scheduler/TestPhasedExecutionSchedule.java [hunk 1]
```diff
@@ -192,7 +192,7 @@ private static PlanFragment createBroadcastJoinPlanFragment(String name, PlanFra
 
         RemoteSourceNode remote = new RemoteSourceNode(new PlanNodeId("build_id"), buildFragment.getId(), ImmutableList.of());
         PlanNode join = new JoinNode(
-                new PlanNodeId(name + "_id"),
+                new PlanNodeId("join" + "_id"),
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/execution/scheduler/TestPhasedExecutionSchedule.java]: this concatenation is a constant now

File: presto-main/src/main/java/com/facebook/presto/sql/planner/EffectivePredicateExtractor.java [hunk 2]
```diff
@@ -215,7 +215,7 @@ public Expression visitWindow(WindowNode node, Void context)
     @Override
     public Expression visitUnion(UnionNode node, Void context)
     {
-        return deriveCommonPredicates(node, source -> node.outputSymbolMap(source).entries());
+        return deriveCommonPredicates(node, source -> node.getMultiSourceSymbolMapping().outputSymbolMap(source).entries());
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/EffectivePredicateExtractor.java]: > The same thing exists in ExchangeNode, I will try to use MultiSourceSymbolMapping there as a follow up of this.

cmt msg doesn't need to contain information about your plans for the future. Consider removing or rephrasing as "`MultiSourceSymbolMapping` can later also be used in `ExchangeNode`."

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/MultiSourceSymbolMapping.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    public MultiSourceSymbolMapping(
+            @JsonProperty("outputToInputs") ListMultimap<Symbol, Symbol> outputToInputs,
+            @JsonProperty("outputs") List<Symbol> outputs,
+            @JsonProperty("sources") List<PlanNode> sources)
+    {
+        this.outputToInputs = ImmutableListMultimap.copyOf(requireNonNull(outputToInputs, "outputToInputs is null"));
+        this.outputs = ImmutableList.copyOf(requireNonNull(outputs, "outputs is null"));
+        this.sources = ImmutableList.copyOf(requireNonNull(sources, "sources is null"));
+
+        checkArgument(!sources.isEmpty(), "Must have at least one source");
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/MultiSourceSymbolMapping.java]: make sure that for each output symbol there is the same number of input symbols (see `ExchangeNode` constructor)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/MultiSourceSymbolMapping.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+    @JsonProperty("sources")
+    public List<PlanNode> getSources()
+    {
+        return sources;
+    }
+
+    public List<Symbol> sourceOutputLayout(int sourceIndex)
+    {
+        // Make sure the sourceOutputLayout symbols are listed in the same order as the corresponding output symbols
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/MultiSourceSymbolMapping.java]: checkState that `sourceIndex < sources.size()`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/MultiSourceSymbolMapping.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+                .map(symbol -> outputToInputs.get(symbol).get(sourceIndex))
+                .collect(toImmutableList());
+    }
+
+    /**
+     * Returns the output to input symbol mapping for the given source channel
+     */
+    public Map<Symbol, SymbolReference> sourceSymbolMap(int sourceIndex)
+    {
+        return outputs.stream()
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/MultiSourceSymbolMapping.java]: ditto

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/MultiSourceSymbolMapping.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+                .collect(toImmutableMap(identity(), symbol -> outputToInputs.get(symbol).get(sourceIndex).toSymbolReference()));
+    }
+
+    /**
+     * Returns the input to output symbol mapping for the given source channel.
+     * A single input symbol can map to multiple output symbols, thus requiring a Multimap.
+     */
+    public Multimap<Symbol, SymbolReference> outputSymbolMap(int sourceIndex)
+    {
+        return Multimaps.transformValues(FluentIterable.from(getOutputSymbols())
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/MultiSourceSymbolMapping.java]: use normal java stream:
```
getOutputSymbols().stream()
  .collect(flatteningToImmutableListMultimap(
    identity(), 
    symbol -> outputToInputs.get(symbol).stream().
      .map(Symbol::toSymbolReference));
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/MultiSourceSymbolMapping.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Function.identity;
+
+@Immutable
+public class MultiSourceSymbolMapping
+{
+    private final ListMultimap<Symbol, Symbol> outputToInputs;
+    private final List<Symbol> outputs;
+    private final List<PlanNode> sources;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/MultiSourceSymbolMapping.java]: I find it strange that this class contains not just the mappings, but also the sources.

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/MultiSourceSymbolMapping.java]: I think I had tried without sources in a first place and then I changed my mind due the validation of symbols in constructor. I will give it a try one more time.

**Pull request closed without merging**